File: src/client/java/com/somefrills/features/update/UpdateManager.java
package com.somefrills.features.update;

import com.google.gson.JsonElement;
import com.somefrills.misc.Utils;
import moe.nea.libautoupdate.CurrentVersion;
import moe.nea.libautoupdate.PotentialUpdate;
import moe.nea.libautoupdate.UpdateContext;
import moe.nea.libautoupdate.UpdateTarget;
import net.fabricmc.loader.api.FabricLoader;

import java.io.IOException;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;

import static com.somefrills.Main.LOGGER;
import static com.somefrills.Main.mc;

public class UpdateManager {
    private static volatile CompletableFuture<?> activePromise = null;
    private static volatile UpdateState updateState = UpdateState.NONE;
    private static volatile PotentialUpdate potentialUpdate = null;
    private static volatile boolean hasCheckedThisSession = false;

    public enum UpdateState {
        AVAILABLE,
        QUEUED,
        DOWNLOADED,
        NONE
    }

    public static UpdateState getUpdateState() {
        return updateState;
    }

    public static String getCurrentVersion() {
        return FabricLoader.getInstance()
                .getModContainer(com.somefrills.Main.MOD_ID)
                .map(mod -> mod.getMetadata().getVersion().getFriendlyString())
                .orElse("unknown");
    }

    public static String getLatestVersion() {
        if (potentialUpdate != null) {
            return potentialUpdate.getUpdate().getVersionNumber().getAsString();
        }
        return getCurrentVersion();
    }

    public static boolean isUpdateAvailable() {
        return updateState == UpdateState.AVAILABLE;
    }

    private static final UpdateContext context = new UpdateContext(
            new CustomGithubReleaseUpdateSource("AverageUser125", "SomeFrills"),
            UpdateTarget.deleteAndSaveInTheSameFolder(UpdateManager.class),
            new CurrentVersion() {
                @Override
                public String display() {
                    return getCurrentVersion();
                }

                @Override
                public boolean isOlderThan(JsonElement element) {
                    if (element == null || !element.isJsonPrimitive()) return false;
                    try {
                        String asString = element.getAsString();
                        int currentParsed = parseSemanticVersion(getCurrentVersion());
                        int latestParsed = parseSemanticVersion(asString);
                        return currentParsed < latestParsed;
                    } catch (Exception e) {
                        LOGGER.warn("Failed to compare versions", e);
                        return false;
                    }
                }
            },
            com.somefrills.Main.MOD_ID
    );

    public static void reset() {
        updateState = UpdateState.NONE;
        potentialUpdate = null;
        hasCheckedThisSession = false;
        cancelActivePromise();
        LOGGER.debug("Update state reset");
    }

    public static void checkUpdate() {
        checkUpdate(false);
    }

    public static void checkUpdate(boolean autoQueue) {
        if (hasCheckedThisSession) {
            LOGGER.info("Already checked for updates this session");
            return;
        }

        if (updateState != UpdateState.NONE && updateState != UpdateState.AVAILABLE) {
            LOGGER.info("Trying to perform update check while another update is already in progress");
            return;
        }

        if (updateState == UpdateState.AVAILABLE) {
            updateState = UpdateState.NONE;
            LOGGER.info("Resetting update state to force download");
        }

        hasCheckedThisSession = true;
        LOGGER.info("Starting update check (autoQueue: {})", autoQueue);
        activePromise = context.checkUpdate("full").thenAcceptAsync(update -> {
            LOGGER.info("Update check completed");
            if (updateState != UpdateState.NONE) {
                LOGGER.info("This appears to be the second update check. Ignoring this one");
                return;
            }

            potentialUpdate = update;
            if (update.isUpdateAvailable()) {
                updateState = UpdateState.AVAILABLE;
                String versionName = update.getUpdate().getVersionName();
                LOGGER.info("Update available: {}", versionName);
                Utils.infoFormat("Update available: {}", versionName);

                if (autoQueue) {
                    LOGGER.info("Auto-queuing update");
                    Utils.infoFormat("Auto-queuing update");
                    queueUpdate();
                }
            } else {
                LOGGER.info("No update available");
            }
        }, mc).exceptionally(e -> {
            LOGGER.error("[SomeFrills] Failed to check for updates", e);
            hasCheckedThisSession = false;
            return null;
        });
    }

    public static void queueUpdate() {
        if (updateState != UpdateState.AVAILABLE) {
            LOGGER.info("Trying to enqueue an update while another one is already downloaded or none is present");
            return;
        }

        if (potentialUpdate == null) {
            LOGGER.error("Cannot queue update: potentialUpdate is null");
            return;
        }

        updateState = UpdateState.QUEUED;
        LOGGER.info("Queuing update download");
        Utils.infoFormat("Queuing update download");
        activePromise = CompletableFuture.supplyAsync(() -> {
            LOGGER.info("Update download started");
            Utils.infoFormat("Update download started");
            try {
                potentialUpdate.prepareUpdate();
            } catch (IOException e) {
                LOGGER.error("Failed to download update", e);
                Utils.infoFormat("Failed to download update: {}", e.getMessage());
                updateState = UpdateState.AVAILABLE;
                hasCheckedThisSession = false;
            } catch (NullPointerException e) {
                LOGGER.error("Update was cleared while downloading", e);
                updateState = UpdateState.AVAILABLE;
                hasCheckedThisSession = false;
            }
            return null;
        }).thenAcceptAsync(__ -> {
            if (potentialUpdate == null) {
                LOGGER.error("Update was cleared before installation");
                return;
            }
            LOGGER.info("Update download completed");
            Utils.infoFormat("Update download completed. Will be installed on next restart.");
            updateState = UpdateState.DOWNLOADED;
            potentialUpdate.executePreparedUpdate();
        }, mc);
    }

    private static void cancelActivePromise() {
        if (activePromise != null) {
            activePromise.cancel(true);
            activePromise = null;
        }
    }

    private static int parseSemanticVersion(String version) {
        if (version == null || version.isEmpty()) return 0;

        String[] numbers = version.split("\\.");
        if (numbers.length < 3) return 0;

        int major = parseInt(numbers[0]).orElse(0);
        int minor = parseInt(numbers[1]).orElse(0);
        int patch = parseInt(numbers[2]).orElse(0);
        return major * 1000 + minor * 100 + patch;
    }

    private static Optional<Integer> parseInt(String str) {
        try {
            return Optional.of(Integer.parseInt(str));
        } catch (NumberFormatException e) {
            return Optional.empty();
        }
    }
}

File: src/main/kotlin/com/somefrills/features/update/UpdateManager.kt
package com.somefrills.features.update

import com.google.gson.JsonElement
import com.somefrills.Main
import com.somefrills.Main.LOGGER
import com.somefrills.Main.mc
import com.somefrills.misc.Utils
import moe.nea.libautoupdate.CurrentVersion
import moe.nea.libautoupdate.PotentialUpdate
import moe.nea.libautoupdate.UpdateContext
import moe.nea.libautoupdate.UpdateTarget
import net.fabricmc.loader.api.FabricLoader
import net.fabricmc.loader.api.ModContainer
import java.io.IOException
import java.util.concurrent.CompletableFuture
import java.util.function.Consumer
import java.util.function.Function
import kotlin.concurrent.Volatile

object UpdateManager {
    private val context = UpdateContext(
        CustomGithubReleaseUpdateSource("AverageUser125", "SomeFrills"),
        UpdateTarget.deleteAndSaveInTheSameFolder(UpdateManager::class.java),
        object : CurrentVersion {
            override fun display(): String {
                return currentVersion
            }

            override fun isOlderThan(element: JsonElement): Boolean {
                if (!element.isJsonPrimitive) return false
                try {
                    val asString = element.asString
                    val currentParsed = parseSemanticVersion(currentVersion)
                    val latestParsed = parseSemanticVersion(asString)
                    return currentParsed < latestParsed
                } catch (e: Exception) {
                    LOGGER.warn("Failed to compare versions", e)
                    return false
                }
            }
        },
        Main.MOD_ID
    )

    @Volatile
    private var activePromise: CompletableFuture<*>? = null

    @Volatile
    var updateState: UpdateState = UpdateState.NONE
        private set

    @Volatile
    private var potentialUpdate: PotentialUpdate? = null

    val currentVersion: String
        get() = FabricLoader.getInstance()
            .getModContainer(Main.MOD_ID)
            .map(Function { mod: ModContainer ->
                mod.metadata.version.friendlyString
            })
            .orElse("unknown") ?: "unknown"

    val latestVersion: String
        get() {
            return potentialUpdate?.update?.getVersionNumber()
                ?.asString ?: currentVersion
        }

    val isUpdateAvailable: Boolean
        get() = updateState == UpdateState.AVAILABLE

    fun reset() {
        updateState = UpdateState.NONE
        potentialUpdate = null
        cancelActivePromise()
        LOGGER.debug("Update state reset")
    }

    @JvmOverloads
    fun checkUpdate(autoQueue: Boolean = false) {
        if (updateState != UpdateState.NONE && updateState != UpdateState.AVAILABLE) {
            LOGGER.info("Trying to perform update check while another update is already in progress")
            return
        }

        if (updateState == UpdateState.AVAILABLE) {
            updateState = UpdateState.NONE
            LOGGER.info("Resetting update state to force download")
        }

        LOGGER.info("Starting update check (autoQueue: {})", autoQueue)
        activePromise = context.checkUpdate("full").thenAcceptAsync(Consumer thenAcceptAsync@{ update: PotentialUpdate? ->
            LOGGER.info("Update check completed")
            if (updateState != UpdateState.NONE) {
                LOGGER.info("This appears to be the second update check. Ignoring this one")
                return@thenAcceptAsync
            }

            potentialUpdate = update
            if (update!!.isUpdateAvailable()) {
                updateState = UpdateState.AVAILABLE
                val versionName = update.update.getVersionName()
                LOGGER.info("Update available: {}", versionName)
                Utils.infoFormat("Update available: {}", versionName)

                if (autoQueue) {
                    LOGGER.info("Auto-queuing update")
                    Utils.infoFormat("Auto-queuing update")
                    queueUpdate()
                }
            } else {
                LOGGER.info("No update available")
            }
        }, mc).exceptionally(Function { e: Throwable? ->
            LOGGER.error("[SomeFrills] Failed to check for updates", e)
            null
        })
    }

    fun queueUpdate() {
        if (updateState != UpdateState.AVAILABLE) {
            LOGGER.info("Trying to enqueue an update while another one is already downloaded or none is present")
            return
        }

        if (potentialUpdate == null) {
            LOGGER.error("Cannot queue update: potentialUpdate is null")
            return
        }

        updateState = UpdateState.QUEUED
        LOGGER.info("Queuing update download")
        Utils.infoFormat("Queuing update download")
        activePromise = CompletableFuture.supplyAsync<Any?> {
            LOGGER.info("Update download started")
            Utils.infoFormat("Update download started")
            try {
                potentialUpdate!!.prepareUpdate()
            } catch (e: IOException) {
                LOGGER.error("Failed to download update", e)
                Utils.infoFormat("Failed to download update: {}", e.message)
                updateState = UpdateState.AVAILABLE
            } catch (e: NullPointerException) {
                LOGGER.error("Update was cleared while downloading", e)
                updateState = UpdateState.AVAILABLE
            }
            null
        }.thenAcceptAsync(Consumer thenAcceptAsync@{ _: Any? ->
            if (potentialUpdate == null) {
                LOGGER.error("Update was cleared before installation")
                return@thenAcceptAsync
            }
            LOGGER.info("Update download completed")
            Utils.infoFormat("Update download completed. Will be installed on next restart.")
            updateState = UpdateState.DOWNLOADED
            potentialUpdate!!.executePreparedUpdate()
        }, mc)
    }

    private fun cancelActivePromise() {
        if (activePromise != null) {
            activePromise!!.cancel(true)
            activePromise = null
        }
    }

    private fun parseSemanticVersion(version: String): Int {
        if (version.isEmpty()) return 0

        val numbers: Array<String> = version.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        if (numbers.size < 3) return 0

        val major = parseInt(numbers[0])
        val minor = parseInt(numbers[1])
        val patch = parseInt(numbers[2])
        return major * 1000 + minor * 100 + patch
    }

    private fun parseInt(str: String): Int {
        return try {
            str.toInt()
        } catch (_: NumberFormatException) {
            0
        }
    }

    enum class UpdateState {
        AVAILABLE,
        QUEUED,
        DOWNLOADED,
        NONE
    }
}
